src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp

```diff
@@ -0,0 +1,239 @@
+/*******************************************************************\
+
+Module: Havoc multiple and possibly dependent targets simultaneously
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+\*******************************************************************/
+
+/// \file
+/// Havoc multiple and possibly dependent targets simultaneously
+
+#include "havoc_assigns_clause_targets.h"
+
+#include <map>
+
+#include <langapi/language_util.h>
+
+#include <util/c_types.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
+#include <util/pointer_expr.h>
+#include <util/pointer_offset_size.h>
+#include <util/pointer_predicates.h>
+#include <util/std_code.h>
+
+#include "utils.h"
+
+static const char ASSIGNS_CLAUSE_REPLACEMENT_TRACKING[] =
+  " assigned by replaced ";
+
+/// Returns a pointer expression to the start address of the target
+///
+/// - If the target is a `pointer_object(p)` expression,
+///   return a pointer to the same object with offset zero
+/// - else, if the target is a sized and assignable expression,
+///   return its address
+/// - else trigger an error.
+///
+static exprt base_pointer(
+  const exprt &target,
+  // context parameters
+  const namespacet &ns)
+{
+  if(target.id() == ID_pointer_object)
+  {
+    const auto &ptr = target.operands().front();
+    // bring the offset back to zero
+    auto res = minus_exprt{ptr, pointer_offset(ptr)};
+    res.type() = ptr.type();
+    return std::move(res);
+  }
+  else if(is_assignable(target))
+  {
+    // check that the type of the target has a definite size
+    const auto &size = size_of_expr(target.type(), ns);
+
+    INVARIANT(
+      size.has_value(),
```

Comment:
Please move this into the `INVARIANT` to avoid a situation where `size` ends up as an unused value (when completely disabling invariants). That is: `INVARIANT(size_of_expr(target.type(), ns).has_value(), ...)`

Reply:
fixed